Draft: Resolve parallel dbt templating issue

alanmcruickshank opened the pull request:
This PR has a couple of things to still shake out. Namely that one test is failing for a reason I don't understand, and it looks like the CLI output is treading on toes when running in parallel. Those issues aside, I think it's ready for a review, at least from an architecture standpoint.

This resolves #1088 .

I've tested it on our large dbt project with the dbt templater and it's FAST.

It also builds on #1103 so will be easier to review once that's merged.

In short:
- We render the sql files in the main thread and then pass them to the child threads for parsing and linting.
- The rest of the changes are making enough methods in the linter _static_ or _class_ methods to do that in a way that we don't have to create a new `Linter` object each time.

alanmcruickshank commented:
The outstanding test is `test__linter__skip_dbt_model_disabled`, which I'm wondering whether this is actually a better solution for. The existing test checks that a `LintedFile` is still returned but that it has a null `templated_file`. In this reworking, no `LintedFile` is returned at all. It's still effectively _skipped_, but skipped in a slightly different way. @NiallRees, @dmateusp  - do you have views on how this might play out for users?

barrywhart commented on src/sqlfluff/core/linter/runner.py:
```diff
@@ -211,3 +207,28 @@ def __init__(self, ee):
+    allow_process_parallelism: bool = True,
+) -> BaseRunner:
+    """Generate a runner instance based on parallel and sytem configuration."""
+    if parallel > 1 and sys.version_info > (3, 7):
```
I'm thinking now that the version check should use `>=`, i.e.: `sys.version_info >= (3, 7)` because it's 3.6.x that is not supporting parallel.

alanmcruickshank replied:
I've updated the comparison. Do you have something I can link to and put in a comment about *why* 3.6 isn't supported. I had a quick google just now and couldn't work it out.

barrywhart replied:
TBH, I don't remember. I know it wasn't working. I recall thinking we could possibly get it working, but 3.6 is so old, was it worth the effort.

I'll see if I can reproduce the error on 3.6 again.

barrywhart commented:
About the CLI output, could we potentially capture the output in the workers and return it to the main process and print it there? That's what I think I did with the linter output in the first parallel implementation.

alanmcruickshank commented:
> About the CLI output, could we potentially capture the output in the workers and return it to the main process and print it there? That's what I think I did with the linter output in the first parallel implementation.

This is what I _intended_ to do here, but evidently didn't manage. I'll have another look.

barrywhart commented:
@alanmcruickshank: I replied to your question about the issue with 3.6. Mentioning it here in case that comment thread gets hidden by GitHub!!

dmateusp commented:
@alanmcruickshank I think not returning a LintedFile makes sense!

alanmcruickshank commented:
I think this is ready to review properly.

The additional changes are:
- Solved the issue with double logging of paths and threaded logging (@barrywhart - I'd just duplicated the code 🤦 - one was inside the thread).
- In debugging this I came across an odd failure with the `fatal` templating errors I added a while back. I've added a counter in the `dbt` templater so that they only actually trigger a fatal error if they happen 3 times in a row. This isn't really related to this PR, but it's a tiny change that was easy to make while I was in the code.
- File names are now explicit fields on `ParsedString` and `RenderedFile`. Previously we introspected these from the contained `TemplatedFile` but in the case of a templating error, this was `None` and so triggered some issues and odd failures.
- There's an additional logging line in the `dbt` templater for the _completion_ of templating so that if you're not running in high verbosity modes it doesn't just look like it's stuck on `templating...` for ages.

alanmcruickshank commented:
I realise the `--bench` command isn't that useful in a parallel world because the timings overlap. I've added a total `Clock time` output to the timing outputs so that we can evaluate performance better.

On 858 files using the dbt templater:
- `-p 1`: Clock time:   418.02
- `-p 4`: Clock time:   261.33

Not quite a 50% reduction, but I'm still pretty happy with that. 🎉

barrywhart commented on src/sqlfluff/core/linter/linter.py:
```diff
@@ -455,13 +461,16 @@ def lint_parsed(
             # than any generated during the fixing cycle.
             violations += initial_linting_errors
+        else:
+            # If no parsed tree, set to Nonw
```
```suggestion
            # If no parsed tree, set to None

```

barrywhart commented on src/sqlfluff/core/templaters/dbt.py:
```diff
@@ -199,12 +206,18 @@ def process(self, *, fname, in_str=None, config=None, formatter=None):
+            return processed_result
         except DbtCompilationException as e:
+            # Increment the counter
+            self._sequential_fails += 1
```
Why would compilation fail the first time but subsequently work?

alanmcruickshank replied:
some compilation fails are fatal (the kind that initially inspired this), but I've realised some aren't. They are where the compilation fails in a file-specific way.

barrywhart reviewed (approved):
A couple of questions, nothing huge. Looks good!

alanmcruickshank commented on src/sqlfluff/cli/commands.py:
```diff
@@ -648,7 +648,7 @@ def parse(path, code_only, format, profiler, bench, nofail, logger=None, **kwarg
                     else None,
                 )
-                for filepath, (parsed, _, _, _, _) in zip(filepaths, result)
+                for filepath, (parsed, _, _, _, _, _) in zip(filepaths, result)
```
Yeah great idea, I think this can be much simpler.

barrywhart commented on src/sqlfluff/cli/commands.py:
```diff
@@ -644,17 +644,16 @@ def parse(path, code_only, format, profiler, bench, nofail, logger=None, **kwarg
                     else None,
                 )
-                for filepath, (parsed, _, _, _, _, _) in zip(filepaths, result)
+                for linted_result in result
```
🎉

barrywhart reviewed (approved):
Very nice!!
